### DISCUSSION TITLE
Disjunction rules exclude members (strings) which contain previous members (substrings)

### DISCUSSION THREAD

**Discussion by lthoerner** [pest-parser/pest#822] [Q&A] [answered]
**Describe the bug**
I'm not sure if I'm using exactly the right terminology for this, so this is a quick vocab key to help the report be more concise.
*Disjunction* - A rule where multiple strings are accepted using the OR (`|`) operator
*Member* - One of the strings in the disjunction
`s1` - An arbitrary string member of a disjunction
`s2` - A string which both occurs after `s1` in the disjunction and starts with `s1`

In a disjunction, if there exist both an `s1` and an `s2`, `s2` will not be an accepted string when applying the rule.

Consider the rule `protocol = { "tcp" | "udp" | "udplite" }`. In this case, `"udplite"` (`s2`) is not considered to be a valid option when parsing, because it contains the substring `"udp"` (`s1`).

__Exceptions:__
- When the rule is defined with the longer string first. For instance, `protocol = { "tcp" | "udplite" | "udp" }` allows `"udplite"` to be recognized.
- When the substring only represents a portion of a previous string. For instance, `protocol = { "tcp" | "udp" | "udlite" }` allows `"udlite"` to be recognized.
- When the substring is not at the beginning of the longer string. For instance, `protocol = { "tcp" | "udp" | "liteudp" }` allows `"liteudp"` to be recognized.

**To Reproduce**
Make any disjunction where a member appears again as a substring at the beginning of a later member.
```
R = { D | s1 ∈ D ∧ (s1 ⧺ z) ∈ D }
```
Where `R` is a rule containing a disjunction `D`, and `s1` and `z` are both arbitrary strings.

**Expected behavior**
It would be expected that, if the input does not obey the `s1` rule, it would then be checked against `s2` before the rule is considered to be violated altogether.

**Additional context**
I am not sure if this is intended behavior. If so, I would like more clarification as to why this is.

**Comment by NoahTheDuke** [comment, marked as answer]
This is a feature of PEG parsers. The OR is called an ordered choice, meaning that each potential options is considered in definition order and the first match is used without checking the rest of the choices. This is important due to PEG's backtracking (and memoization), which allows for arbitrarily nested rules without ambiguity.
